amoro-ams/src/main/java/org/apache/amoro/server/optimizing/maintainer/IcebergTableMaintainer.java

```diff
@@ -696,26 +702,152 @@ CloseableIterable<FileEntry> fileScan(
+              Comparable<Object> partitionValue =
+                  contentFile.partition().get(entry.getKey(), entry.getValue().getClass());
+              boolean expired = partitionValue.compareTo(entry.getValue()) < 0;
+              expiredList.add(expired);
+            }
+            if (!expiredList.isEmpty() && expiredList.stream().allMatch(Boolean::booleanValue)) {
+              expiredFiles.addFile(fileEntry);
+            }
+          });
+      return true;
+    }
+    return false;
+  }
+
+  private void expireByMetricsUpperBound(
+      CloseableIterable<FileEntry> entries,
+      DataExpirationConfig expirationConfig,
+      long expireTimestamp,
+      ExpireFiles expiredFiles) {
+    Map<StructLike, DataFileFreshness> partitionFreshness = Maps.newConcurrentMap();
+    Queue<FileEntry> fileEntries = new LinkedTransferQueue<>();
+    entries.forEach(
+        e -> {
+          if (mayExpired(e, partitionFreshness, expireTimestamp)) {
+            fileEntries.add(e);
+          }
+        });
+    fileEntries
+        .parallelStream()
+        .filter(e -> willNotRetain(e, expirationConfig, partitionFreshness))
+        .forEach(expiredFiles::addFile);
+  }
+
+  private Map<Integer, Map<Integer, PartitionField>> buildExpirePartitionFieldsMap(
+      Types.NestedField expireField) {
+    Map<Integer, Map<Integer, PartitionField>> partitionFieldsMap = new HashMap<>();
+    for (Map.Entry<Integer, PartitionSpec> entry : table.specs().entrySet()) {
+      int pos = 0;
+      Map<Integer, PartitionField> posToField = new HashMap<>();
+      for (PartitionField field : entry.getValue().fields()) {
+        if (field.sourceId() == expireField.fieldId()) {
+          posToField.put(pos, field);
+        }
+        pos++;
+      }
+      partitionFieldsMap.put(entry.getKey(), posToField);
+    }
+
+    return partitionFieldsMap;
+  }
+
+  private Map<Integer, Map<Integer, Comparable<?>>> getExpirePartitionValueMap(
+      Map<Integer, Map<Integer, PartitionField>> expirePartitionFieldsMap,
+      Types.NestedField field,
+      Comparable<?> expireValue) {
+    Map<Integer, Map<Integer, Comparable<?>>> expirePartitionValue = new HashMap<>();
+    for (Map.Entry<Integer, Map<Integer, PartitionField>> entry :
+        expirePartitionFieldsMap.entrySet()) {
+      Map<Integer, Comparable<?>> posToValue = new HashMap<>();
+      for (Map.Entry<Integer, PartitionField> posToField : entry.getValue().entrySet()) {
+        posToValue.put(
+            posToField.getKey(),
+            ((SerializableFunction<Comparable<?>, Comparable<?>>)
+                    posToField.getValue().transform().bind(field.type()))
+                .apply(expireValue));
+      }
+      expirePartitionValue.put(entry.getKey(), posToValue);
+    }
+    return expirePartitionValue;
+  }
+
+  private Comparable<?> getPartitionUpperBound(
+      DataExpirationConfig expirationConfig, Types.NestedField field, long expireTimestamp) {
+    switch (field.type().typeId()) {
```

Comment:
could you please add some comments on this?  especially for the type `TIMESAMPT` and `LONG`